File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/Lang/SpanishLanguagePack.swift
import Foundation

private func chars(_ labels: String...) -> [KeyDef] {
    labels.map { KeyDef($0, Int($0.unicodeScalars.first!.value)) }
}

public struct SpanishLanguagePack: LanguagePack {
    public let id = "es"
    public let displayName = "Español"
    public let locale = Locale(identifier: "es")

    public init() {}

    public var alphaRows: [[KeyDef]] {
        [
            chars("q", "w", "e", "r", "t", "y", "u", "i", "o", "p"),
            chars("a", "s", "d", "f", "g", "h", "j", "k", "l", "ñ"),
            [KeyDef("⬆", SpecialKeys.shift, widthWeight: 1.5)]
                + chars("z", "x", "c", "v", "b", "n", "m")
                + [KeyDef("←", SpecialKeys.backspace, widthWeight: 1.5)],
            [
                KeyDef("?123", SpecialKeys.symbols, widthWeight: 1.5),
                KeyDef("🌐", SpecialKeys.globe, widthWeight: 1.0),
                KeyDef(",", Int(Character(",").unicodeScalars.first!.value), widthWeight: 1.0),
                KeyDef(" ", Int(Character(" ").unicodeScalars.first!.value), widthWeight: 4.0),
                KeyDef(".", Int(Character(".").unicodeScalars.first!.value), widthWeight: 1.0),
                KeyDef("↵", SpecialKeys.enter, widthWeight: 1.5),
            ],
        ]
    }

    public var symbols1Rows: [[KeyDef]] { EnglishLanguagePack().symbols1Rows }
    public var symbols2Rows: [[KeyDef]] { EnglishLanguagePack().symbols2Rows }

    public let longPressAccents: [Character: [Character]] = [
        "a": ["á", "à", "ä", "â", "ã"],
        "e": ["é", "è", "ê", "ë"],
        "i": ["í", "ì", "î", "ï"],
        "o": ["ó", "ò", "ô", "ö", "õ"],
        "u": ["ú", "ù", "û", "ü"],
        "?": ["¿"],
        "!": ["¡"],
    ]
}

File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/lang/SpanishLanguagePack.kt
package com.draftright.keyboard.lang

import com.draftright.keyboard.KeyDef
import com.draftright.keyboard.LanguagePack
import com.draftright.keyboard.SpecialKeys
import java.util.Locale

private fun chars(vararg labels: String): List<KeyDef> =
    labels.map { KeyDef(it, it[0].code) }

object SpanishLanguagePack : LanguagePack {
    override val id = "es"
    override val displayName = "Español"
    override val locale: Locale = Locale("es")

    override val alphaRows: List<List<KeyDef>> = listOf(
        chars("q", "w", "e", "r", "t", "y", "u", "i", "o", "p"),
        chars("a", "s", "d", "f", "g", "h", "j", "k", "l", "ñ"),
        listOf(KeyDef("⬆", SpecialKeys.SHIFT, 1.5f)) +
            chars("z", "x", "c", "v", "b", "n", "m") +
            listOf(KeyDef("←", SpecialKeys.BACKSPACE, 1.5f)),
        listOf(
            KeyDef("?123", SpecialKeys.SYMBOLS, 1.5f),
            KeyDef("🌐", SpecialKeys.GLOBE, 1.0f),
            KeyDef(",", ','.code, 1.0f),
            KeyDef(" ", ' '.code, 4.0f),
            KeyDef(".", '.'.code, 1.0f),
            KeyDef("↵", SpecialKeys.ENTER, 1.5f),
        ),
    )

    override val symbols1Rows: List<List<KeyDef>> = EnglishLanguagePack.symbols1Rows
    override val symbols2Rows: List<List<KeyDef>> = EnglishLanguagePack.symbols2Rows

    override val longPressAccents: Map<Char, List<Char>> = mapOf(
        'a' to listOf('á', 'à', 'ä', 'â', 'ã'),
        'e' to listOf('é', 'è', 'ê', 'ë'),
        'i' to listOf('í', 'ì', 'î', 'ï'),
        'o' to listOf('ó', 'ò', 'ô', 'ö', 'õ'),
        'u' to listOf('ú', 'ù', 'û', 'ü'),
        '?' to listOf('¿'),
        '!' to listOf('¡'),
    )
}
